README.md

# gchemol

gchemol is a graph-based chemical object library implemented in Rust programming
language.

[![Build Status](https://travis-ci.org/ybyygu/gchemol.svg?branch=master)](https://travis-ci.org/ybyygu/gchemol)
[![GPL3 licensed](https://img.shields.io/badge/license-GPL3-blue.svg)](./LICENSE)
[![Built with Spacemacs](https://cdn.rawgit.com/syl20bnr/spacemacs/442d025779da2f62fc86c2082703697714db6514/assets/spacemacs-badge.svg)](http://spacemacs.org)


# Features

-   Fast and safe.
-   Easy to deploy in server environment.
-   core graph data structure using [petgraph](https://github.com/bluss/petgraph)
-   read molecules in various formats using [nom](https://github.com/Geal/nom) parser combinators
-   linear algebra backed by [nalgebra](http://nalgebra.org/)
-   render molecule in user defined formats by templating with [handlebars](https://github.com/sunng87/handlebars-rust)


# How to use


## install rust

follow the official guide:

-   [Installation · The Rust Programming Language](https://www.rust-lang.org/en-US/install.html)


## setup

add gchemol dependency to your Cargo.toml:

    [dependencies]
    gchemol = {git = "https://github.com/ybyygu/gchemol"}


# Edit molecule


## atom

    use gchemol::Atom;
    
    // construct from element and position
    let a = Atom::new("C", [0.0, 0.0, 0.0]);
    let b = Atom::new("C", [1.2, 1.2, 1.2]);

or simply convert from a string:

    let a: Atom = "C 1.0 1.0 0.2"
        .parse()
        .expect("atom from string");

set more attributes using the builder pattern

    let a = Atom::build()
        .symbol("Fe")
        .position(1.2, 1.0, 0.3)
        .finish();


## molecule

1.  Creating a molecule manually

        use gchemol::Molecule;
        
        let atoms = [
            Atom::new("C", [ 0.000000,   0.000000,  0.000000]),
            Atom::new("C", [ 0.000000,   0.000000,  1.089000]),
            Atom::new("C", [ 1.026719,   0.000000, -0.363000]),
            Atom::new("C", [-0.513360,  -0.889165, -0.363000]),
            Atom::new("C", [-0.513360,   0.889165, -0.363000])];
        
        // create a molecule named methane
        let mut mol = Molecule::new("methane");
        // add atoms in a loop
        for a in atoms {
            mol.add_atom(a);
        }

2.  Reading and writing molecules

        use gchemol::io;
        use gchemol::Molecule;
        use gchemol::prelude::*;
        
        // Read an xyz file and write to a Gaussian Input file.
        let mol = Molecule::from_file("path/to/file").unwrap();
        mol.to_file("methane.gjf")
        
        // get the total number of atoms
        let na = mol.natoms();
        // get the total number of bonds
        let nb = mol.nbonds();
        
        // read multiple molecules (trajectory) from a chemical file
        // the number of atoms in different frame could be different
        let mols = io::read("path/to/trajectory.xyz").unwrap();

3.  Coordinates

        let mut positions = mol.positions();
        positions[0] = [1.2, 1.0, 0.1];
        mol.set_positions(positions);

4.  Sorted molecule

    create new molecule with all atoms sorted by element number (hydrogen last):
    
        let m = mol.sorted();
    
    to be continued &#x2026;


## bonds

1.  bonding connectivity

        let b = mol.get_bond(bond_index);
        let (a1, a2) = b.partners(&mol);
        
        let (atom_index1, atom_index2) = mol.partners(bond_index);
        
        let neighbors = mol.neighbors(atom_index);
    
    TBD: more &#x2026;


## lattice

build a periodic structure

    use gchemol::Lattice;
    
    // build lattice from lattice parameters
    let lattice = Lattice::from_params
        (
            3.84,                   // a
            3.84,                   // b
            3.84,                   // c
            120.,                   // alpha
            90.,                    // beta
            60.,                    // gamma
        );
    
    // build lattice from cell vectors
    let lat = Lattice::new([
        [ 15.3643,   0.    ,   0.    ], // vector a
        [  4.5807,  15.5026,   0.    ], // vector b
        [  0.    ,   0.    ,  17.4858], // vector c
    ]);
    
    // set periodic boundary conditions for molecule
    mol.set_lattice(lat);
    
    // remove periodic boundary conditions
    mol.unbuild_crystal();


# Alignment

    use geoemtry::Alignment;
    
    // superpose using a subset of structures
    let reference = positions1[0..5];
    let candidate = positions1[0..5];
    
    // align candidate onto reference
    let mut align = Alignment::new(&candidate);
    let sp = align.superpose(&reference, None).unwrap();
    
    // apply superposition to all atoms
    let new = sp.apply(&candidate);


# Templating

For syntax details: [sunng87/handlebars-rust: Rust templating with Handlebars](https://github.com/sunng87/handlebars-rust)

molecule title

    {{molecule_title}}

total number of atoms in molecule

    {{molecule.number_of_atoms}}

total number of bonds in molecule

    {{molecule.number_of_bonds}}

loop over atoms:

    {{#each molecule.atoms as |a| ~}}
    {{a.index}} {{a.symbol}} {{a.x}} {{a.y}} {{a.z}}
    {{/each~}}

Atom index (a.index) is counting from 1.

element types (Fe 4, C 5 &#x2026;):

    {{#each molecule.element_types as |e| ~}} {{e.0}} {{e.1}} {{/each}}

elment index of atom (element index in element types array, counting from 1):

    {{a.element_index}}

atom properties

    {{atom.symbol}}

cartesian coordinates

    {{atom.x}} {{atom.y}} {{atom.z}}

fractional coordinates

    {{atom.fx}} {{atom.fy}} {{atom.fz}}

unit cell parameters

    {{molecule.unit_cell.a}}
    {{molecule.unit_cell.b}}
    {{molecule.unit_cell.c}}
    {{molecule.unit_cell.alpha}}
    {{molecule.unit_cell.beta}}
    {{molecule.unit_cell.gamma}}
    {{molecule.unit_cell.va}}
    {{molecule.unit_cell.vb}}
    {{molecule.unit_cell.vc}}

format string or number:

    {{format 12.16 width=12 prec=6 align="right"}}

alignment control:

    {{align="left"}}
    {{align="right"}}
    {{align="center"}}


# Related projects

-   lumol
-   ase
-   pymatgen


# References

-   [Lessons from sixteen years of molecular simulation in Python | Konrad Hinsen's Blog](https://khinsen.wordpress.com/2013/04/10/lessons-from-sixteen-years-of-molecular-simulation-in-python/)



Changelog.md

# v0.0.32 - <span class="timestamp-wrapper"><span class="timestamp">[2019-06-02 Sun]</span></span>

-   fixed compiling errors in new version of rust (1.37)


# v0.0.31 - <span class="timestamp-wrapper"><span class="timestamp">[2019-04-29 Mon]</span></span>

-   update nalgebra to v0.18
-   serialize/deserialize \`Molecule\` using serde


# v0.0.30 - <span class="timestamp-wrapper"><span class="timestamp">[2019-04-17 Wed]</span></span>

-   fixed compiling error with new version of Rust
-   io: public \`describe\_backends\` method


# v0.0.29 - <span class="timestamp-wrapper"><span class="timestamp">[2019-04-08 Mon]</span></span>

-   add a struct for making supercell: \`Supercell\`


# v0.0.28 - <span class="timestamp-wrapper"><span class="timestamp">[2019-04-06 Sat]</span></span>

-   add \`fgt\` helper for rendering molecule
-   add \`species\` data for rendering molecule
-   fix issues caused by incorrect uses of \`relative\_eq\` macro


# v0.0.27 - <span class="timestamp-wrapper"><span class="timestamp">[2018-12-20 Thu]</span></span>

-   update quicli to v0.4
-   update rand to v0.6


# v0.0.26 - <span class="timestamp-wrapper"><span class="timestamp">[2018-11-24 Sat]</span></span>

-   add tests for geometry mod
-   fixed compiling errors caused by rand mod



src/lattice.rs
// [[file:~/Workspace/Programming/gchemol/gchemol.note::891f59cf-3963-4dbe-a7d2-48279723b72e][891f59cf-3963-4dbe-a7d2-48279723b72e]]
//===============================================================================#
//   DESCRIPTION:  Represents 3D periodic lattices
//
//       OPTIONS:  ---
//  REQUIREMENTS:  ---
//         NOTES:  ---
//        AUTHOR:  Wenping Guo <ybyygu@gmail.com>
//       LICENCE:  GPL version 3
//       CREATED:  <2018-04-29 14:27>
//       UPDATED:  <2018-08-06 Mon 11:35>
//===============================================================================#

use nalgebra::{
    Vector3,               // A stack-allocated, 3-dimensional column vector.
    Matrix3,               // A stack-allocated, column-major, 3x3 square matrix
};

type Mat3D = Matrix3<f64>;
type Vec3D = Vector3<f64>;
use quicli::prelude::*;
// 891f59cf-3963-4dbe-a7d2-48279723b72e ends here

// [[file:~/Workspace/Programming/gchemol/gchemol.note::b17e625d-352f-419e-9d10-a84fcdb9ff07][b17e625d-352f-419e-9d10-a84fcdb9ff07]]
/// Periodic 3D lattice
#[derive(Debug, Clone, Copy)]
pub struct Lattice {
    /// internal translation matrix
    matrix: Mat3D,
    /// Lattice origin
    origin: Vec3D,

    /// Cached inverse of lattice matrix
    inv_matrix: Option<Mat3D>,

    /// Cached volume of the unit cell.
    volume: Option<f64>,

    /// The perpendicular widths of the unit cell on each direction,
    /// i.e. the distance between opposite faces of the unit cell
    widths: Option<[f64; 3]>,

    /// Cached cell lengths parameters
    lengths: Option<[f64; 3]>,

    /// Cached cell angles parameters
    angles: Option<[f64; 3]>,
}

// matrix inversion
fn get_inv_matrix(matrix: Mat3D) -> Mat3D {
    matrix.try_inverse().expect("bad matrix")
}

// cell volume
fn get_cell_volume(mat: Mat3D) -> f64 {
    let va = mat.column(0);
    let vb = mat.column(1);
    let vc = mat.column(2);
    va.dot(&vb.cross(&vc))
}

// return cell length parameters
fn get_cell_lengths(mat: Mat3D) -> [f64; 3] {
    [
        mat.column(0).norm(),
        mat.column(1).norm(),
        mat.column(2).norm()
    ]
}

// return cell angle parameters in degrees
fn get_cell_angles(mat: Mat3D) -> [f64; 3] {
    let va = mat.column(0);
    let vb = mat.column(1);
    let vc = mat.column(2);
    [
        vb.angle(&vc).to_degrees(),
        va.angle(&vc).to_degrees(),
        va.angle(&vb).to_degrees(),
    ]
}

impl Default for Lattice {
    fn default() -> Self {
        Lattice {
            matrix: Mat3D::identity(),
            origin: Vec3D::zeros(),

            inv_matrix: None,
            volume: None,
            widths: None,
            lengths: None,
            angles: None,
        }
    }
}

impl Lattice {
    pub fn new<T: Into<[[f64; 3]; 3]>>(tvs: T) -> Self {
        Lattice {
            matrix: Mat3D::from(tvs.into()),
            ..Default::default()
        }
    }

    /// using a cache to reduce the expensive matrix inversion calculations
    fn inv_matrix(&mut self) -> Mat3D {
        // make a readonly reference
        let matrix = self.matrix;
        let im = self.inv_matrix.get_or_insert_with(|| matrix.try_inverse().expect("bad matrix"));

        *im
    }

    fn get_cell_widths(&mut self)  -> [f64; 3] {
        let volume = self.volume();
        let [van, vbn, vcn] = self.lengths();

        let wa = volume / (vbn*vcn);
        let wb = volume / (vcn*van);
        let wc = volume / (van*vbn);

        [wa, wb, wc]
    }

    pub fn widths(&mut self) -> [f64; 3] {
        if let Some(ws) = self.widths {
            return ws;
        } else {
            let ws = self.get_cell_widths();
            self.widths = Some(ws);

            ws
        }
    }

    /// Return the volume of the unit cell
    /// the cache will be updated if necessary
    pub fn volume(&mut self) -> f64 {
        // make a read-only reference
        let mat = self.matrix;
        let volume = self.volume.get_or_insert_with(|| get_cell_volume(mat));

        *volume
    }

    /// Construct lattice from lattice parameters
    /// Unit cell angles in degrees, lengths in Angstrom
    pub fn from_params(a: f64, b: f64, c: f64, alpha: f64, beta: f64, gamma: f64) -> Self {
        let alpha = alpha.to_radians();
        let beta  = beta.to_radians();
        let gamma = gamma.to_radians();

        let acos = alpha.cos();
        let bcos = beta.cos();
        let gcos = gamma.cos();
        let gsin = gamma.sin();
        let v = (1.
                 - acos.powi(2)
                 - bcos.powi(2)
                 - gcos.powi(2)
                 + 2.0 * acos * bcos * gcos).sqrt();

        let va = [a,
                  0.0,
                  0.0];

        let vb = [b*gcos,
                  b*gsin,
                  0.0
                  ];

        let vc = [c*bcos,
                  c*(acos - bcos*gcos)/gsin,
                  c*v/gsin];

        Lattice::new([va, vb, vc])
    }

    /// Set cell origin in Cartesian coordinates
    pub fn set_origin(&mut self, loc: [f64; 3]) {
        self.origin = Vec3D::from(loc);
    }

    /// Lattice length parameters: a, b, c
    pub fn lengths(&mut self) -> [f64; 3] {
        let mat = self.matrix;
        let lengths = self.lengths.get_or_insert_with(|| get_cell_lengths(mat));

        [
            lengths[0],
            lengths[1],
            lengths[2],
        ]
    }

    /// Lattice angle parameters in degrees
    pub fn angles(&mut self) -> [f64; 3] {
        let mat = self.matrix;
        let angles = self.angles.get_or_insert_with(|| get_cell_angles(mat));

        [
            angles[0],
            angles[1],
            angles[2],
        ]
    }

    // FIXME: cell widths
    /// Scale Lattice by a positive constant
    pub fn scale_by(&mut self, v: f64) {
        debug_assert!(v > 0.);
        self.matrix *= v;

        // reset caches
        self.inv_matrix = None;
        self.volume = None;
        self.widths = None;
        self.lengths = None;
        self.angles = None;
    }

    /// Get cell origin in Cartesian coordinates
    pub fn origin(&self) -> [f64; 3] {
        self.origin.into()
    }

    /// Returns the fractional coordinates given cartesian coordinates.
    pub fn to_frac(&mut self, p: [f64; 3]) -> [f64; 3] {
        let im = self.inv_matrix();
        let v = Vec3D::from(p);
        let fs = im * (v - self.origin);
        fs.into()
    }

    /// Returns the cartesian coordinates given fractional coordinates.
    pub fn to_cart(&self, p: [f64; 3]) -> [f64; 3] {
        let v = Vec3D::from(p);
        let fs = self.matrix*v + self.origin;

        fs.into()
    }

    /// Lattice vector a
    pub fn vector_a(&self) -> [f64; 3] {
        self.matrix.column(0).transpose().into()
    }

    /// Lattice vector b
    pub fn vector_b(&self) -> [f64; 3] {
        self.matrix.column(1).transpose().into()
    }

    /// Lattice vector c
    pub fn vector_c(&self) -> [f64; 3] {
        self.matrix.column(2).transpose().into()
    }

    /// Lattice vectors
    pub fn vectors(&self) -> [[f64; 3]; 3] {
        self.matrix.into()
    }

    /// Check if lattice is orthorhombic
    pub fn is_orthorhombic(&self) -> bool {
        let diag = self.matrix.diagonal();
        let m = Matrix3::from_diagonal(&diag);
        m == self.matrix
    }
}
// b17e625d-352f-419e-9d10-a84fcdb9ff07 ends here

// [[file:~/Workspace/Programming/gchemol/gchemol.note::83cff231-cc63-4077-b07e-a26a2c2b906d][83cff231-cc63-4077-b07e-a26a2c2b906d]]
use std::f64;

impl Lattice {
    /// Return the relevant periodic images required for neighborhood search
    /// within cutoff radius
    pub fn relevant_images(&mut self, radius: f64) -> Vec<Vector3<f64>> {
        let ns = self.n_min_images(radius);
        let na = ns[0] as isize;
        let nb = ns[1] as isize;
        let nc = ns[2] as isize;

        let mut images = vec![];
        for i in (-na)..(na+1) {
            for j in (-nb)..(nb+1) {
                for k in (-nc)..(nc+1) {
                    let v = Vec3D::from([i as f64, j as f64, k as f64]);
                    images.push(v);
                }
            }
        }

        images
    }

    /// Return the minimal number of images for neighborhood search on each cell
    /// direction within cutoff radius
    fn n_min_images(&mut self, radius: f64) -> [usize; 3]{
        let mut ns = [0; 3];

        for (i, &w) in self.widths().iter().enumerate() {
            let n = (radius / w).ceil();
            ns[i] = n as usize;
        }

        ns
    }

    /// Return the distance between two points computed using the minimum image
    /// convention.
    ///
    /// Reference
    /// ---------
    /// (1) Tuckerman, M. E. Statistical Mechanics: Theory and Molecular
    /// Simulation, 1 edition.; Oxford University Press: Oxford ; New York,
    /// 2010.
    fn distance_tuckerman(&mut self, pi: [f64; 3], pj: [f64; 3]) -> f64 {
        let pij = [pj[0] - pi[0],
                   pj[1] - pi[1],
                   pj[2] - pi[2]];

        // apply minimum image convention on the scaled coordinates
        let mut fij = self.to_frac(pij);
        for x in 0..3 {
            fij[x] -= fij[x].round();
        }

        // transform back to cartesian coordinates
        let pij = self.to_cart(fij);
        Vec3D::from(pij).norm()
    }

    /// Return the shortest distance between `pi` (point i) and the periodic
    /// images of `pj` (point j). This algorithm will loop over all relevant
    /// images
    fn distance_brute_force(&mut self, pi: [f64; 3], pj: [f64; 3]) -> f64 {
        // The cutoff radius for finding relevant images.
        // Use the value from Tuckerman algorithm as cutoff radius, since it is
        // always larger than the real distance using minimum image convention
        let cutoff = self.distance_tuckerman(pi, pj);
        let relevant_images = self.relevant_images(cutoff);

        let mut distance = f64::MAX;
        for v in relevant_images {
            let dd = self.to_cart(v.into());
            let ipij = [
                pj[0] + dd[0] - pi[0],
                pj[1] + dd[1] - pi[1],
                pj[2] + dd[2] - pi[2],
            ];

            let d = Vec3D::from(ipij).norm();
            if d < distance {
                distance = d;
            }
        }

        distance
    }

    // TODO: return the closest periodic image?
    /// Return the shortest distance between `pi` (point i) and the periodic
    /// images of `pj` (point j) under the minimum image convention
    pub fn distance(&mut self, pi: [f64; 3], pj: [f64; 3]) -> f64 {
        if self.is_orthorhombic() {
            self.distance_tuckerman(pi, pj)
        } else {
            self.distance_brute_force(pi, pj)
        }
    }
}
// 83cff231-cc63-4077-b07e-a26a2c2b906d ends here

// [[file:~/Workspace/Programming/gchemol/gchemol.note::5754ca07-a93d-47e5-8256-c7236777b2ee][5754ca07-a93d-47e5-8256-c7236777b2ee]]
use molecule::Molecule;

impl Molecule {
    /// Set periodic lattice
    pub fn set_lattice(&mut self, lat: Lattice) {
        self.lattice = Some(lat);
    }

    /// Unbuild current crystal structure leaving a nonperiodic structure
    pub fn unbuild_crystal(&mut self) {
        self.lattice = None
    }

    /// Return fractional coordinates relative to unit cell.
    pub fn scaled_positions(&self) -> Option<Vec<[f64; 3]>> {
        if let Some(mut lat) = self.lattice {
            let mut fxyzs = vec![];
            for a in self.atoms() {
                let xyz = a.position();
                let fxyz = lat.to_frac(xyz);
                fxyzs.push(fxyz)
            }
            Some(fxyzs)
        } else {
            None
        }
    }

    /// Set positions relative to unit cell.
    pub fn set_scaled_positions(&mut self, scaled: &[[f64; 3]]) -> Result<()> {
        if let Some(mut lat) = self.lattice {
            let mut positions = vec![];
            for &p in scaled {
                let xyz = lat.to_cart(p);
                positions.push(p);
            }

            self.set_positions(positions)
        } else {
            bail!("cannot set scaled positions for aperiodic structure")
        }
    }
}
// 5754ca07-a93d-47e5-8256-c7236777b2ee ends here

// [[file:~/Workspace/Programming/gchemol/gchemol.note::4bc21235-f285-4976-a32a-b33506381b58][4bc21235-f285-4976-a32a-b33506381b58]]
#[test]
fn test_lattice_construct() {
    let mut lat = Lattice::default();
    let loc = [1.0, 2.0, 3.0];
    lat.set_origin(loc);
    assert_eq!(loc, lat.origin());

    let mut lat = Lattice::new([[ 15.3643,   0.    ,   0.    ],
                                [  4.5807,  15.5026,   0.    ],
                                [  0.    ,   0.    ,  17.4858]]);

    let [a, b, c] = lat.lengths();
    assert_eq!(false, lat.is_orthorhombic());

    assert_relative_eq!(a, 15.3643, epsilon=1e-4);
    assert_relative_eq!(b, 16.1652, epsilon=1e-4);
    assert_relative_eq!(c, 17.4858, epsilon=1e-4);

    let [alpha, beta, gamma] = lat.angles();
    assert_relative_eq!(alpha, 90.0, epsilon=1e-4);
    assert_relative_eq!(beta, 90.0, epsilon=1e-4);
    assert_relative_eq!(gamma, 73.5386, epsilon=1e-4);

    let mut lat = Lattice::from_params(a, b, c, alpha, beta, gamma);
    assert_eq!([a, b, c], lat.lengths());
    assert_eq!([alpha, beta, gamma], lat.angles());
}

#[test]
fn test_lattice_neighborhood() {
    let mut lat = Lattice::new([
        [ 18.256,   0.   ,   0.   ],
        [  0.   ,  20.534,   0.   ],
        [  0.   ,   0.   ,  15.084],
    ]);
    assert_eq!(true, lat.is_orthorhombic());

    assert_eq!([1, 1, 1], lat.n_min_images(9.));
    assert_eq!([2, 1, 2], lat.n_min_images(19.));
    assert_eq!([2, 1, 2], lat.n_min_images(20.));
    assert_eq!([2, 2, 2], lat.n_min_images(20.6));

    let expected = [
        Vec3D::new(-1.0, -1.0, -1.0),
        Vec3D::new(-1.0, -1.0,  0.0),
        Vec3D::new(-1.0, -1.0,  1.0),
        Vec3D::new(-1.0,  0.0, -1.0),
        Vec3D::new(-1.0,  0.0,  0.0),
        Vec3D::new(-1.0,  0.0,  1.0),
        Vec3D::new(-1.0,  1.0, -1.0),
        Vec3D::new(-1.0,  1.0,  0.0),
        Vec3D::new(-1.0,  1.0,  1.0),
        Vec3D::new( 0.0, -1.0, -1.0),
        Vec3D::new( 0.0, -1.0,  0.0),
        Vec3D::new( 0.0, -1.0,  1.0),
        Vec3D::new( 0.0,  0.0, -1.0),
        Vec3D::new( 0.0,  0.0,  0.0),
        Vec3D::new( 0.0,  0.0,  1.0),
        Vec3D::new( 0.0,  1.0, -1.0),
        Vec3D::new( 0.0,  1.0,  0.0),
        Vec3D::new( 0.0,  1.0,  1.0),
        Vec3D::new( 1.0, -1.0, -1.0),
        Vec3D::new( 1.0, -1.0,  0.0),
        Vec3D::new( 1.0, -1.0,  1.0),
        Vec3D::new( 1.0,  0.0, -1.0),
        Vec3D::new( 1.0,  0.0,  0.0),
        Vec3D::new( 1.0,  0.0,  1.0),
        Vec3D::new( 1.0,  1.0, -1.0),
        Vec3D::new( 1.0,  1.0,  0.0),
        Vec3D::new( 1.0,  1.0,  1.0)];

    let images = lat.relevant_images(3.0);
    assert_eq!(expected.len(), images.len());
    assert_eq!(expected[1][2], images[1][2]);
}

#[test]
fn test_lattice_volume() {
    let vts = [
        [ 5.,  0.,  0.],
        [ 5.,  5.,  0.],
        [ 1.,  0.,  5.]
    ];

    let mut lat = Lattice::new(vts);
    assert_eq!(vts, lat.vectors());
    assert_eq!(vts[0], lat.vector_a());
    assert_eq!(vts[1], lat.vector_b());
    assert_eq!(vts[2], lat.vector_c());

    assert_relative_eq!(125.0, lat.volume(), epsilon=1e-4);
    lat.scale_by(4.);
    assert_relative_eq!(8000.0, lat.volume(), epsilon=1e-4);
}

#[test]
fn test_lattice_frac_cart() {
    // ovito/tests/files/LAMMPS/multi_sequence_1.dump
    let mut lat = Lattice::new([[5.09, 0.00, 0.00],
                                [0.00, 6.74, 0.00],
                                [0.00, 0.00, 4.53]]);

    let fs = lat.to_frac([2.1832, 1.6850, 3.8505]);
    assert_relative_eq!(fs[0], 0.4289, epsilon=1e-3);
    assert_relative_eq!(fs[1], 0.2500, epsilon=1e-3);
    assert_relative_eq!(fs[2], 0.8500, epsilon=1e-3);
    let fs = lat.to_frac([6.9068, 5.0550, 0.6795]);
    assert_relative_eq!(fs[0], 1.3569, epsilon=1e-3);
    let fs = lat.to_frac([4.3618, 5.0550, 1.5855]);
    assert_relative_eq!(fs[2], 0.3500, epsilon=1e-3);

    let coords = lat.to_cart([0.4289, 0.2500, 0.8500]);
    assert_relative_eq!(coords[0], 2.1832, epsilon=1e-3);
    assert_relative_eq!(coords[1], 1.6850, epsilon=1e-3);
    assert_relative_eq!(coords[2], 3.8505, epsilon=1e-3);
}

#[test]
fn test_lattice_mic_distance() {
    let mut lat = Lattice::new([
        [5.0, 0.0, 0.0],
        [1.0, 5.0, 0.0],
        [1.0, 1.0, 5.0],
    ]);

    // the shortest distance: 2.61383
    let d = lat.distance_tuckerman([0.; 3], [-0.94112, -4.34823, 2.53058]);
    assert_relative_eq!(2.66552, d, epsilon=1e-4);
    let d = lat.distance_brute_force([0.; 3], [-0.94112, -4.34823, 2.53058]);
    assert_relative_eq!(2.61383, d, epsilon=1e-4);

    // the shortest distance: 2.53575
    let d = lat.distance_tuckerman([0.; 3], [-2.46763, 0.57717, 0.08775]);
    assert_relative_eq!(2.59879, d, epsilon=1e-4);
    let d = lat.distance_brute_force([0.; 3], [-2.46763, 0.57717, 0.08775]);
    assert_relative_eq!(2.53575, d, epsilon=1e-4);
}
// 4bc21235-f285-4976-a32a-b33506381b58 ends here
